frame/bags-list/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+// limitations under the License.
+
+//! # Bags-List Pallet
+//!
+//! A semi-sorted list, where items hold an `AccountId` based on some `VoteWeight`. The `AccountId`
+//! (`id` for short) might be synonym to a `voter` or `nominator` in some context, and `VoteWeight`
+//! signifies the chance of each id being included in the final [`VoteWeightProvider::iter`].
+//!
+//! It implements [`sp_election_provider_support::SortedListProvider`] to provide a semi-sorted list
+//! of accounts to another pallet. It needs some other pallet to give it some information about the
+//! weights of accounts via [`sp_election_provider_support::VoteWeightProvider`].
+//!
+//! This pallet is not configurable at genesis. Whoever uses it should call appropriate functions of
+//! the `SortedListProvider` (e.g. `on_insert`, or `regenerate`) at their genesis.
+//!
+//! # Goals
+//!
+//! The data structure exposed by this pallet aims to be optimized for:
+//!
+//! - insertions and removals.
+//! - iteration over the top* N items by weight, where the precise ordering of items doesn't
+//!   particularly matter.
+//!
+//! # Details
+//!
+//! - items are kept in bags, which are delineated by their range of weight (See [`BagThresholds`]).
+//! - for iteration, bags are chained together from highest to lowest and elements within the bag
+//!   are iterated from head to tail.
+//! - items within a bag are iterated in order of insertion. Thus removing an item and re-inserting
+//!   it will worsen its position in list iteration; this reduces incentives for some types of spam
+//!   that involve consistently removing and inserting for better position. Further, ordering
+//!   granularity is thus dictated by range between each bag threshold.
+//! - if an item's weight changes to a value no longer within the range of its current bag the
+//!   item's position will need to be updated by an external actor with rebag (update), or removal
+//!   and insertion.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_election_provider_support::{SortedListProvider, VoteWeight, VoteWeightProvider};
+use frame_system::ensure_signed;
+use sp_std::prelude::*;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+mod list;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+pub use list::Error;
+use list::List;
+
+pub(crate) const LOG_TARGET: &'static str = "runtime::bags_list";
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("[{:?}] 👜", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	pub struct Pallet<T>(_);
```

Comment:
can we already add `generate_storage_info`?

Reply:
104775681571805d9183c6440b4afe735eb8092b